internal/component/discovery/discovery.go

```diff
@@ -225,10 +225,19 @@ func (c *Component) runDiscovery(ctx context.Context, d DiscovererWithMetrics) {
 				send()
 				haveUpdates = false
 			}
-		case <-ctx.Done():
+		// If Run has exited, return.
+		// We can't exit straight after the context is cancelled.
+		// We need to receive from runExited first.
+		// If we don't the first "go func() " will block on sending to runExited.
+		case <-runExited:
```

Comment:
The [CI was failing](https://drone.grafana.net/grafana/alloy/2364/2/2) because of this. I didn't see this problem when running tests locally though - it's a flaky test.